url89849c8c5b_timmolter_XChart_tgz-pJ8-org_knowm_xchart_standalone_Example1J8/src/org/knowm/xchart/XChartPanel.java

```diff
@@ -0,0 +1,364 @@
+import javax.swing.JFileChooser;
+import javax.swing.JMenuItem;
+import javax.swing.JPanel;
+import javax.swing.JPopupMenu;
+import javax.swing.KeyStroke;
+import javax.swing.filechooser.FileFilter;
+
+import org.knowm.xchart.BitmapEncoder.BitmapFormat;
+import org.knowm.xchart.VectorGraphicsEncoder.VectorGraphicsFormat;
+import org.knowm.xchart.internal.chartpart.Chart;
+
+/**
+ * A Swing JPanel that contains a Chart
+ * <p>
+ * Right-click + Save As... or ctrl+S pops up a Save As dialog box for saving the chart as PNG, JPEG, etc. file.
+ *
+ * @author timmolter
+ */
+public class XChartPanel<T extends Chart> extends JPanel {
+
+  private final T chart;
+  private final Dimension preferredSize;
+  private String saveAsString = "Save As...";
+  private String exportAsString = "Export As...";
+
+  /**
+   * Constructor
+   *
+   * @param chart
+   */
+  public XChartPanel(final T chart) {
+
+    this.chart = chart;
+    preferredSize = new Dimension(chart.getWidth(), chart.getHeight());
+
+    // Right-click listener for saving chart
+    this.addMouseListener(new PopUpMenuClickListener());
+
+    // Control+S key listener for saving chart
+    KeyStroke ctrlS = KeyStroke.getKeyStroke(KeyEvent.VK_S, Toolkit.getDefaultToolkit().getMenuShortcutKeyMask());
+    this.getInputMap(WHEN_IN_FOCUSED_WINDOW).put(ctrlS, "save");
+    this.getActionMap().put("save", new SaveAction());
+
+    // Control+E key listener for saving chart
+    KeyStroke ctrlE = KeyStroke.getKeyStroke(KeyEvent.VK_E, Toolkit.getDefaultToolkit().getMenuShortcutKeyMask());
+    this.getInputMap(WHEN_IN_FOCUSED_WINDOW).put(ctrlE, "export");
+    this.getActionMap().put("export", new ExportAction());
+  }
+
+  /**
+   * Set the "Save As..." String if you want to localize it.
+   *
+   * @param saveAsString
+   */
+  public void setSaveAsString(String saveAsString) {
+
+    this.saveAsString = saveAsString;
+  }
+
+  /**
+   * Set the "Export As..." String if you want to localize it.
+   *
+   * @param exportAsString
+   */
+  public void setExportAsString(String exportAsString) {
+
+    this.exportAsString = exportAsString;
+  }
+
+  @Override
+  protected void paintComponent(Graphics g) {
+
+    super.paintComponent(g);
+
+    Graphics2D g2d = (Graphics2D) g.create();
+    chart.paint(g2d, getWidth(), getHeight());
+    g2d.dispose();
+  }
+
+  public T getChart() {
+
+    return this.chart;
+  }
+
+  @Override
+  public Dimension getPreferredSize() {
+
+    return preferredSize;
+  }
+
+  private class SaveAction extends AbstractAction {
+
+    public SaveAction() {
+
+      super("save");
+    }
+
+    @Override
+    public void actionPerformed(ActionEvent e) {
+
+      showSaveAsDialog();
+    }
+  }
+
+  private class ExportAction extends AbstractAction {
+
+    public ExportAction() {
+
+      super("export");
+    }
+
+    @Override
+    public void actionPerformed(ActionEvent e) {
+
+      showExportAsDialog();
+    }
+  }
+
+  private void showSaveAsDialog() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 153-172 into a separate method.